### ISSUE TITLE
Upgrading the plugin to MathJax v3

### ISSUE THREAD

**Issue by liffiton** [liffiton/dokuwiki-plugin-mathjax#37]
Pull request #36 contributes some changes that make the plugin work with MathJax v3 instead of v2.  v3 has a slightly different way to insert MathJax into a page, and it's not compatible with v2.  To upgrade the plugin cleanly, then, we have to be careful about not breaking things for any existing users with customized setups using v2.

Auto-detecting the version feels way too fragile, and it's not feasible in some cases (e.g., self-hosting the libraries with no version in the path).  Adding a configuration option to choose between the two versions sounds nice, but the default value should be to use 3.x, and that would make any existing plugin users have that setting.  I don't know how to add a new configuration option to the plugin with a default that *only* applies for a new install and not an upgrade...

One option would be to create a new configuration option like `'url_v3'` or similar and set its default to be a CDN URL for Mathjax v3.  Then set the existing `'url'` option's default to be an empty string.  The plugin could do v2-compatible things if the existing `'url'` option is set (meaning the user customized it); otherwise, do v3-compatible things.  I think the same thing would have to be done for the `'config'` option as well.  This ends up being unwieldy for the common case, though, with confusing leftover configuration options that aren't used in most cases.

If there are no good options here, it might be best to just release a `mathjax3` plugin separately and direct people to it instead of this one.  I don't see any critical improvements in v3 for the use case of this plugin, so letting existing installs linger on v2 doesn't cause any harm.  (Correct me if I'm wrong, but the only v3 change I see that's relevant to the use here is that it renders faster.)

Ideas are welcome.

**Comment by greneholt** [comment]
You could use a combination of options. You could try to auto-detect the version from the URL, and if that fails, fall back to a configuration setting that would default to use V2.

So there would be two options: the URL, which defaults to V3, and a select box that asks "MathJax Version to use if URL-based auto-detection fails", which defaults to V2.

If you make the version detection conservative, this should avoid breaking any existing users, while still supporting locally hosted files. One way to do the version detection would be if the URL ends in `MathJax.js?config=*`, it is v2, whereas if it ends in one of the following, it is v3:

- tex-chtml.js
- tex-chtml-full.js
- tex-svg.js
- tex-svg-full.js
- tex-mml-chtml.js
- tex-mml-svg.js.
- mml-chtml.js.
- mml-svg.js
- startup.js

**Comment by Ragos81** [comment]
As this is an old post:
For my understanding, there is no plan to switch to v3 and I will not waste time by using v2 when switching from mathpublish to MathJax (no judgement, no offense, just curiosity)?

**Comment by liffiton** [comment]
I'd be happy to get a pull request to review, but I am not actively developing this plugin these days.  I haven't had a compelling need myself to move to v3.

There are a few forks of this repository where people have updated it for Mathjax3: [1](https://github.com/gkzhb/dokuwiki-plugin-mathjax3), [2](https://github.com/bennvan/dokuwiki-plugin-mathjax), [3](https://github.com/greneholt/dokuwiki-plugin-mathjax).  I haven't tried them myself.

**Comment by Ragos81** [comment]
Totally get your point, never change a running system!

**Comment by marmat123** [comment]
> To upgrade the plugin cleanly, then, we have to be careful about not breaking things for any existing users with customized setups using v2.

Yes, so the idea would be to write a new plugin, even if it's in name only, and say that the change only affects the JavaScript library used and it's not wise to have both enabled or to replace one with the other without checking.

So separate what doesn't go together.

Then anyone who wants to can use the new plugin. Advantage: anyone who sets up a new page can use the new one straight away, otherwise everyone can update as they wish. Both are unproblematic for the developer.

Over time, the old plugin will be phased out.

**Comment by liffiton** [comment]
Yup, I agree.

Anyone could try one of the forks I linked above that aim to update this plugin to use Mathjax 3.  No one has registered a Mathjax 3 plugin with the [Dokuwiki plugins list](https://www.dokuwiki.org/plugins), though, that I can see.

Alternatively, the [Katex plugin](https://www.dokuwiki.org/plugin:katex) is another way to get the same functionality from a more recent JS library.  Lots of folks prefer Katex to Mathjax anyway.

**Comment by Ragos81** [comment]
@liffiton 
I'm using the [fork](https://github.com/bennvan/dokuwiki-plugin-mathjax) from @bennvan since around 6 month, so far without problems.

**Comment by liffiton** [comment]
> [@liffiton](https://github.com/liffiton) I'm using the [fork](https://github.com/bennvan/dokuwiki-plugin-mathjax) from [@bennvan](https://github.com/bennvan) since around 6 month, so far without problems.

Awesome.  Thanks for the report!
